### PR TITLE
Train/test splitting for dataloaders

### PR DESCRIPTION
**Pull request by MissingNO57**

### CODE REVIEW

File: libs/ml/include/ml/dataloaders/dataloader.hpp [hunk 1]
```diff
@@ -50,13 +50,19 @@ class DataLoader
 
   virtual ~DataLoader() = default;
 
-  virtual ReturnType GetNext() = 0;
+  virtual ReturnType GetNext(bool is_validation = false) = 0;
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/dataloader.hpp]: rather than pass a bool into every function, it seems more logical to have a mode flag for the dataloader, i.e.:

enum class DataLoaderMode
{
    TRAIN,
    VALIDATE,
    TEST,
    INVALID
} 

class DataLoader
{

private:
  DataLoaderMode mode_ = DataLoaderMode::INVALID;
}

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/dataloaders/dataloader.hpp [hunk 2]
```diff
@@ -44,8 +52,9 @@ class DataLoader
    * @param label_shape
    * @param data_shapes
    */
-  explicit DataLoader(bool random_mode)
+  explicit DataLoader(bool random_mode, DataLoaderMode mode)
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/dataloader.hpp]: I think DataLoader should have defaults for both modes:
random = false,
DataLoaderMode = TRAIN

Comment (baykaner) [reply libs/ml/include/ml/dataloaders/dataloader.hpp]: This is because it's completely valid to use a dataloader without being aware of either of these features

File: libs/ml/include/ml/dataloaders/dataloader.hpp [hunk 3]
```diff
@@ -54,15 +63,18 @@ class DataLoader
 
   virtual ReturnType PrepareBatch(fetch::math::SizeType subset_size, bool &is_done_set);
 
-  virtual std::uint64_t Size() const   = 0;
-  virtual bool          IsDone() const = 0;
-  virtual void          Reset()        = 0;
+  virtual std::uint64_t Size() const        = 0;
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/dataloader.hpp]: SizeType

File: libs/ml/include/ml/dataloaders/dataloader.hpp (hunk 2 above)
Comment (baykaner) [reply libs/ml/include/ml/dataloaders/dataloader.hpp]: Oh I see that there is a default constructor and default initialisations so I guess this is fine if you just get rid of DataLoaderMode::INVALID, and set default to TRAIN

File: libs/ml/include/ml/dataloaders/commodity_dataloader.hpp [hunk 4]
```diff
@@ -44,15 +44,16 @@ class CommodityDataLoader : public DataLoader<LabelType, InputType>
...
 
-  ReturnType GetNext() override;
-  SizeType   Size() const override;
-  bool       IsDone() const override;
-  void       Reset() override;
+  ReturnType  GetNext() override;
+  SizeType    Size() const override;
+  bool        IsDone() const override;
+  void        Reset() override;
+  inline bool IsValidable() const override;
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/commodity_dataloader.hpp]: i'm not convinced this method should exist.

It seems like the only place it is used is by the estimators to check whether to compute validation loss. However, it seems to me that anything you do with an estimator you MUST be able to validate (or else the estimator should throw when you try to pass it a dataloader that doesn't handle that behaviour).

File: libs/ml/include/ml/dataloaders/dataloader.hpp [hunk 5]
```diff
@@ -54,15 +63,18 @@ class DataLoader
...
+  virtual bool          IsValidable() const = 0;
+  void                  SetMode(DataLoaderMode new_mode);
 
   template <typename X, typename D>
   friend struct fetch::serializers::MapSerializer;
 
 protected:
-  bool random_mode_ = false;
+  bool           random_mode_ = false;
+  DataLoaderMode mode_        = DataLoaderMode::INVALID;
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/dataloader.hpp]: default to train

(dataloader behaviour & interface should not change - or only minimally - from current implementation if you choose to ignore the new train/val/test split feature)

File: libs/ml/include/ml/dataloaders/tensor_dataloader.hpp [hunk 6]
```diff
@@ -99,12 +107,19 @@ TensorDataLoader<LabelType, InputType>::GetNext()
   {
     throw std::runtime_error("random mode not implemented for tensor dataloader");
   }
+
+  if (this->mode_ == DataLoaderMode::VALIDATE)
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/tensor_dataloader.hpp]: a single if/else is not very expensive, but it is happening inside of the GetNext() method, which is very time critical. We can leave this optimisation for another PR, but I'll point it out here:
Instead of SetMode(new_mode) switching the enum, and then having this if/else based on the enum, we could instead have a third cursor which is a pointer to either the validation or training cursor - Set mode would update the enum and make this third cursor point to either the validation or training data as necessary. Then inside GetNext() we have shorter code and no if/else statement


One way we could make these even faster is if when SetMode(new_mode) is called,

File: libs/ml/include/ml/dataloaders/tensor_dataloader.hpp [hunk 7]
```diff
@@ -113,38 +128,76 @@ template <typename LabelType, typename InputType>
...
+  {
+    return n_train_samples_;
+  }
 }
 
 template <typename LabelType, typename InputType>
 bool TensorDataLoader<LabelType, InputType>::IsDone() const
 {
-  return (data_cursor_ >= data_.shape(batch_data_dim_));
+  if (this->mode_ == DataLoaderMode::TRAIN)
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/tensor_dataloader.hpp]: similarly as with GetNext() we would not need the if/else statement here either.

If you do not implement the suggested change, please raise an issue and leave a TODO in the GetNext

File: libs/ml/include/ml/dataloaders/tensor_dataloader.hpp [hunk 8]
```diff
@@ -113,38 +128,76 @@ template <typename LabelType, typename InputType>
...
+    throw std::runtime_error("Other modes than TRAIN and VALIDATE not supported.");
+  }
 }
 
 template <typename LabelType, typename InputType>
 void TensorDataLoader<LabelType, InputType>::Reset()
 {
-  data_cursor_  = 0;
-  label_cursor_ = 0;
+  if (this->mode_ == DataLoaderMode::TRAIN)
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/tensor_dataloader.hpp]: same again

File: libs/ml/include/ml/dataloaders/tensor_dataloader.hpp [hunk 9]
```diff
@@ -42,24 +42,29 @@ class TensorDataLoader : public DataLoader<LabelType, InputType>
 public:
   TensorDataLoader() = default;
   TensorDataLoader(SizeVector const &label_shape, std::vector<SizeVector> const &data_shapes,
-                   bool random_mode = false);
+                   bool random_mode = false, float VALIDATION_TO_TRAIN_RATIO = 0.0);
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/tensor_dataloader.hpp]: now I think about it, maybe both of these parameters (random & validation ratio) should instead be part of AddData?

what do you think?

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/dataloaders/tensor_dataloader.hpp (hunk 9 above)
Comment (MissingNO57) [reply libs/ml/include/ml/dataloaders/tensor_dataloader.hpp]: I am not convinced, maybe we can add a setter for that.

File: libs/ml/include/ml/dataloaders/word2vec_loaders/w2v_dataloader.hpp [hunk 10]
```diff
@@ -434,6 +451,15 @@ std::vector<std::string> W2VLoader<T>::PreprocessString(std::string const &s)
   return words;
 }
 
+template <typename T>
+void W2VLoader<T>::UpdateCursor()
+{
+  if (this->mode_ != DataLoaderMode::TRAIN)
+  {
+    throw std::runtime_error("Other mode than training not supported yet.");
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/word2vec_loaders/w2v_dataloader.hpp]: It's not a case of 'yet' - it will never be supported for this dataloader.

File: libs/ml/include/ml/dataloaders/tensor_dataloader.hpp [hunk 11]
```diff
@@ -127,20 +133,92 @@ template <typename LabelType, typename InputType>
...
+  *test_cursor_       = test_offset_;
+  *validation_cursor_ = validation_offset_;
+
+  UpdateCursor();
+}
+
+template <typename LabelType, typename InputType>
+void TensorDataLoader<LabelType, InputType>::UpdateCursor()
+{
+  if (this->mode_ == DataLoaderMode::TRAIN)
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/tensor_dataloader.hpp]: switch case is more appropriate than if/else here

File: libs/ml/examples/mnist_estimator/main.cpp [hunk 12]
```diff
@@ -57,9 +57,12 @@ int main(int ac, char **av)
   estimator_config.print_stats    = true;
 
   // setup dataloader
-  auto       data_loader_ptr = std::make_shared<DataLoaderType>(av[1], av[2]);
-  TensorType test_label      = (data_loader_ptr->GetNext()).first;
-  TensorType test_input      = (data_loader_ptr->GetNext()).second.at(0);
+  auto data_loader_ptr = std::make_shared<DataLoaderType>(av[1], av[2]);
+  data_loader_ptr->SetTestRatio(0.2f);
+
+  auto       test_data  = data_loader_ptr->GetNext();
```
Comment (baykaner) [inline libs/ml/examples/mnist_estimator/main.cpp]: i think this gets a piece of training data right?
In any case, wouldn't it be better to use your new functionality instead of doing this and call Validate?

File: libs/ml/include/ml/estimators/dnn_classifier.hpp [hunk 13]
```diff
@@ -139,8 +140,10 @@ bool DNNClassifier<TensorType>::Train(SizeType n_steps)
 template <typename TensorType>
 bool DNNClassifier<TensorType>::Train(SizeType n_steps, DataType &loss)
 {
+  data_loader_ptr_->SetMode(dataloaders::DataLoaderMode::TRAIN);
+
   loss = DataType{0};
-  DataType min_loss;
+  DataType min_loss, val_loss = fetch::math::numeric_max<DataType>();
```
Comment (baykaner) [inline libs/ml/include/ml/estimators/dnn_classifier.hpp]: style, prefer:
DataType min_loss = fetch::math::numeric_max<>();
DataType val_loss = fetch::math::numeric_max<>();

File: libs/ml/include/ml/estimators/dnn_classifier.hpp [hunk 14]
```diff
@@ -187,9 +192,20 @@ bool DNNClassifier<TensorType>::Train(SizeType n_steps, DataType &loss)
 }
 
 template <typename TensorType>
-bool DNNClassifier<TensorType>::Validate()
+bool DNNClassifier<TensorType>::Validate(DataType &validation_loss)
 {
-  throw std::runtime_error("validate not implemented");
+  data_loader_ptr_->SetMode(dataloaders::DataLoaderMode::TEST);
```
Comment (baykaner) [inline libs/ml/include/ml/estimators/dnn_classifier.hpp:200]: slightly confusing - the Validate call has mode TEST?

File: libs/ml/include/ml/estimators/dnn_classifier.hpp [hunk 15]
```diff
@@ -203,9 +219,17 @@ bool DNNClassifier<TensorType>::Predict(TensorType &input, TensorType &output)
 }
 
 template <typename TensorType>
-void DNNClassifier<TensorType>::PrintStats(SizeType epoch, DataType loss)
+void DNNClassifier<TensorType>::PrintStats(SizeType epoch, DataType loss, DataType val_loss)
 {
-  std::cout << "epoch: " << epoch << ", loss: " << loss << std::endl;
+  if (val_loss == fetch::math::numeric_max<DataType>())
```
Comment (baykaner) [inline libs/ml/include/ml/estimators/dnn_classifier.hpp]: wouldn't it be better for this if/else to be a switch/case based on current mode_ enum?

File: libs/ml/include/ml/dataloaders/mnist_loaders/mnist_loader.hpp [hunk 16]
```diff
@@ -147,8 +183,8 @@ class MNISTLoader : public DataLoader<LabelType, InputType>
   void SetupWithDataFiles(std::string const &images_file, std::string const &labels_file)
   {
     std::uint32_t record_length(0);
-    data_   = ReadMnistImages(images_file, size_, record_length);
-    labels_ = read_mnist_labels(labels_file, size_);
+    data_   = ReadMnistImages(images_file, total_size_, record_length);
+    labels_ = read_mnist_labels(labels_file, total_size_);
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/mnist_loaders/mnist_loader.hpp]: style ReadMNistLabels

File: libs/ml/tests/ml/dataloader/tensor_dataloader.cpp [hunk 17]
```diff
@@ -60,6 +60,8 @@ TYPED_TEST(TensorDataloaderTest, serialize_tensor_dataloader)
   // initialise a new tensor dataloader with the wrong shape parameters
   // these will get updated by deserialisation
   fetch::ml::dataloaders::TensorDataLoader<TypeParam, TypeParam> tdl_2({1, 1}, {{1, 1}});
+  tdl_2.SetTestRatio(0.5);
```
Comment (baykaner) [inline libs/ml/tests/ml/dataloader/tensor_dataloader.cpp:63]: what happens if the dataset is a single example?
Does it have a validation and test size of 0 as intended?

maybe we should cover this with a test?

File: libs/ml/include/ml/dataloaders/commodity_dataloader.hpp [hunk 18]
```diff
@@ -152,6 +170,15 @@ void CommodityDataLoader<LabelType, InputType>::GetAtIndex(CommodityDataLoader::
   buffer_.second = std::vector<InputType>({data_.View(index).Copy()});
 }
 
+template <typename LabelType, typename InputType>
+void CommodityDataLoader<LabelType, InputType>::UpdateCursor()
+{
+  if (this->mode_ != DataLoaderMode::TRAIN)
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/commodity_dataloader.hpp]: might want to double check this with emma - seems like it would be trivial to have a validation support for commodity dataloader

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/dataloaders/tensor_dataloader.hpp [hunk 19]
```diff
@@ -186,15 +210,16 @@ void TensorDataLoader<LabelType, InputType>::UpdateRanges()
 template <typename LabelType, typename InputType>
 void TensorDataLoader<LabelType, InputType>::UpdateCursor()
 {
-  if (this->mode_ == DataLoaderMode::TRAIN)
+  switch (this->mode_)
   {
+  case DataLoaderMode::TRAIN:
     this->current_cursor_ = train_cursor_;
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/tensor_dataloader.hpp:217]: style: put cases inside { } also

**Review by baykaner — CHANGES_REQUESTED** [review]
just one style change

**Review by baykaner — DISMISSED** [review]

**Review by baykaner — DISMISSED** [review]

**Review by baykaner — APPROVED** [review]
